File: trunk/development/GSAdmin/GSAdmin/BroadcastMessage.vb
Imports System.Windows.Forms
Imports GameServerInfo
Imports Rcon

Public Class BroadcastMessage

    Public serverinfo As New Collection

    Private Sub OK_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OK_Button.Click
        Try


            If chkList.CheckedItems.Count = 0 Then
                MsgBox("You must select a server to send a message to. Or click close")
                GoTo retry
            End If
            Dim response As String
            Dim i = 0
            Dim s = 0
            For Each item In chkList.CheckedItems
                While i < My.Settings.servers.Count
                    Dim records = My.Settings.servers(i)
                    If records.Contains(item) Then
                        Dim rcon As New Rcon.RCON
                        Dim record = Split(My.Settings.servers(i), "^")
                        response = rcon.sendCommand("say " & TextBox1.Text, record(0), record(1), record(2))
                        s = s + 1
                        GoTo nexty
                    End If

                    i = i + 1

                End While

nexty:
            Next item
            If response.Contains("print") Then
                MsgBox("Message successfully broadcast to " & s & " servers.")
            End If
        Catch ex As Exception

        End Try
retry:

    End Sub

    Private Sub Cancel_Button_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Cancel_Button.Click
        chkList.Items.Clear()
        Me.DialogResult = System.Windows.Forms.DialogResult.Cancel
        Me.Close()
    End Sub

    Private Sub BroadcastMessage_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Try
            Dim counter
            Dim records
            Dim record
            Dim password
            records = My.Settings.servers.Count
            counter = 0
            chkList.Items.Clear()

            While counter < records

                record = Split(My.Settings.servers(counter), "^")
                Dim server As New GameServer(record(0), record(2), GameType.Quake3, 500)
                Dim serverstring
                server.QueryServer()
                If server.IsOnline Then
                    If My.Settings.servers(counter).Contains(GameServer.CleanName(server.Name)) Then
                        password = Split(My.Settings.servers(counter), "^")
                    End If
                    chkList.Items.Add(GameServer.CleanName(server.Name))
                    serverstring = record(0) & "^" & record(2) & "^" & password(1)

                Else
                End If

                counter = counter + 1
            End While

        Catch ex As Exception

        End Try

    End Sub


    Private Sub Button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button1.Click
        Try
            For i = 0 To chkList.Items.Count - 1
                chkList.SetItemChecked(i, True)
            Next i
            OK_Button.Enabled = True

        Catch ex As Exception

        End Try

    End Sub

    Private Sub Button2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button2.Click
        Try
            For i = 0 To chkList.Items.Count - 1
                chkList.SetItemChecked(i, False)
            Next i
            OK_Button.Enabled = False
        Catch ex As Exception

        End Try
    End Sub

    Private Sub chkList_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles chkList.MouseUp
        If chkList.CheckedItems.Count = 0 Then
            OK_Button.Enabled = False
        Else
            OK_Button.Enabled = True
        End If
    End Sub
End Class

File: development/GSAdmin/GSAdminC/broadcastMessage.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace GSAdminC
{
    public partial class broadcastMessage : Form
    {
        Properties.Settings settings = new Properties.Settings();
        Functions func = new Functions();

        public broadcastMessage()
        {
            InitializeComponent();
        }

        private void broadcastMessage_Load(object sender, EventArgs e)
        {
            try
            {
                int records = settings.servers.Count;
                int counter = 0;
                string record;
                string[] field;
                chkList.Items.Clear();

                while (counter < records)
                {
                    record = settings.servers[counter];
                    field = record.Split('^');
                    GameServerInfo.GameServer server = new GameServerInfo.GameServer(field[0], Convert.ToInt32(field[1]), GameServerInfo.GameType.Quake3);
                   
                    server.QueryServer();
                    
                    if (server.IsOnline)
                    {                  
                    chkList.Items.Add(GameServerInfo.GameServer.CleanName(server.Name));
                    }                    
                    counter++;   
                }

                #region listview enabling
                if (chkList.Items.Count > 0)
                {
                    chkList.Enabled = true;
                    Button1.Enabled = true;
                    Button2.Enabled = true;
                    OK_Button.Enabled = true;
                    
                }
                else
                {
                    chkList.Enabled = false;
                    Button1.Enabled = false;
                    Button2.Enabled = false;
                    OK_Button.Enabled = false;
                }
                #endregion
                
            }
            catch (Exception ex)
            {
                func.exceptionbox(ex.Message);  
            }   
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            for (int i = 0; i < chkList.Items.Count; i++)
            {
                chkList.SetItemChecked(i, true);
            }  
        }

        private void Button2_Click(object sender, EventArgs e)
        {
            for (int i = 0; i < chkList.Items.Count; i++)
            {
                chkList.SetItemChecked(i, false);
            }
        }

        private void Cancel_Button_Click(object sender, EventArgs e)
        {

        }

        private void OK_Button_Click(object sender, EventArgs e)
        {
            if (chkList.CheckedItems.Count == 0)
            {
                throw new Exception("You must select at least one server to broadcast to. Alternatively, click close");
            }
            int i = 0;
            string search;
            string[] field;
            int online = 0;
            int offline = 0;
            string response;
            foreach (string item in chkList.CheckedItems)
            {
                i = 0;
                while (i < settings.servers.Count)
                {
                    search = settings.servers[i];
                    if (search.Contains(item))
                    {
                        RCON rcon = new RCON();
                        field = search.Split('^');
                        rcon.initServer(field[0], Convert.ToInt32(field[1]), field[2]);
                        response = rcon.sendCommand("say " + TextBox1.Text);
                        if (func.cmdValid(response))
                        {
                            online++;
                        }
                        else
                        {
                            offline++;
                        }
                        i++;
                    }else{
                        i++;
                    }

                }
            }
            if (offline != 0)
            {
                func.successbox("Message sent to " + Convert.ToString(online) + " of " + Convert.ToString(chkList.CheckedItems.Count) + " servers. " + Environment.NewLine + Convert.ToString(offline) + " server(s) reported errors. Please check your rcon passwords, and ensure the servers are online.");
            }
            else
            {
                func.successbox("Message broadcast to " + Convert.ToString(online) + " servers.");
            }

        }
    }
}
